Cython/Compiler/ExprNodes.py

```diff
@@ -2876,6 +2876,8 @@ def infer_type(self, env, iterator_type=None):
             iterator_type = self.iterator.infer_type(env)
         if iterator_type.is_ptr or iterator_type.is_array:
             return iterator_type.base_type
+        elif iterator_type is py_object_type:
+            return py_object_type
```

Comment:
I think we must look at `iterator_type.sequence` here.
I'm ok with hacking this in for 0.29.x right in `NextNode`, but only specifically for `bytearray_type`. The master branch has a proper fix.

Reply:
The relevant part of the parse tree after `AnalyseExpressionsTransform` without the patch:
```python
- stats[1]: ForInStatNode(pos=(test.py:3:4))
  - target: NameNode(type=<CNumericType unsigned char>, name="elem")
  - item: NextNode(type=<CNumericType unsigned char>)
  - iterator: IteratorNode(type=<PyObjectType>)
    - sequence: NameNode(type=<PyByteArray_Type>, name="data")
  - body: StatListNode(pos=(test.py:4:8))
    - stats[0]: PrintStatNode(pos=(test.py:4:8))
      - arg_tuple: TupleNode(type=<PyTuple_Type>)
        - args[0]: CoerceToPyTypeNode(type=<PyObjectType>)
          - arg: NameNode(type=<CNumericType unsigned char>, name="elem")
```
The IteratorNode is `py_object_type`, should it be `bytearray_type`? The code suggests that `IteratorNode` implements `result = iter(sequence)`, when I do that in python I get a bytearray_iterator.
```python
>>> iter(bytearray(b'1234'))
<bytearray_iterator object at 0x7f1f941729a0>
```
  
Replacing the condition with `elif self.iterator.sequence.type is bytearray_type:` also seems to pass the tests, would that be better?

Reply:
> Replacing the condition with elif self.iterator.sequence.type is bytearray_type: also seems to pass the tests, would that be better?

Yes, I think so.

---------------------

> The master branch has a proper fix.

I kind of wonder if this is a general problem with any type that has a known, optimized index return type but no optimized for loop. Although at the moment the only example we know of is `bytearray`. I'm not sure whether we should worry about it or not - the fix in the master branch to optimize `bytearray` should make it very easy to optimize the loop for other types like this.

I'd probably go for the fix targetted specifically a `bytearray`, but I'm not completely convinced that this won't pop up again for some other type.